xyzpy/gen/batch.py

```diff
@@ -1373,18 +1404,154 @@ def qsub_grow(
+    return gen_cluster_script(crop, scheduler, batch_ids=batch_ids, **kwargs)
+
+
+def qsub_grow(
+    crop, batch_ids=None, *, scheduler='sge',
+    **kwargs
+):  # pragma: no cover
+    """Automagically submit SGE or PBS jobs to grow all missing results.
+    Deprecated in favour of `grow_cluster` and will be removed in the future.
+
+    Parameters
+    ----------
+    crop : Crop
+        The crop to grow.
+    batch_ids : int or tuple[int]
+        Which batch numbers to grow, defaults to all missing batches.
+    scheduler : {'sge', 'pbs'}, optional
+        Whether to use a SGE or PBS submission script template.
+    kwargs
+        See `grow_cluster` for all other parameters.
+    """
+    warnings.warn("'qsub_grow' is deprecated in favour of "
+                  "`grow_cluster` and will be removed in the future",
+                  DeprecationWarning)
+    grow_cluster(crop, scheduler, batch_ids=batch_ids, **kwargs)
+
+
 Crop.gen_qsub_script = gen_qsub_script
 Crop.qsub_grow = qsub_grow
+Crop.gen_cluster_script = gen_cluster_script
+Crop.grow_cluster = grow_cluster
+
+_specific_scheduler_gen_cluster_script_doc = \
+    """Generate a {scheduler} script to grow a Crop.
+
+    Parameters
+    ----------
+    crop : Crop
+        The crop to grow.
+    batch_ids : int or tuple[int]
+        Which batch numbers to grow, defaults to all missing batches.
+    kwargs
+        See `gen_cluster_script` for all other parameters (except for
+        `scheduler` which is not needed here).
+    """
+
+
+def gen_sge_script(crop, batch_ids=None, **kwargs):  # pragma: no cover
+    return gen_cluster_script(crop, 'sge', batch_ids=batch_ids, **kwargs)
+
+
+gen_sge_script.__doc__ = \
+    _specific_scheduler_gen_cluster_script_doc.format(scheduler='SGE')
+
+
+def gen_pbs_script(crop, batch_ids=None, **kwargs):  # pragma: no cover
+    return gen_cluster_script(crop, 'pbs', batch_ids=batch_ids, **kwargs)
+
+
+gen_pbs_script.__doc__ = \
+    _specific_scheduler_gen_cluster_script_doc.format(scheduler='PBS')
+
+
+def gen_slurm_script(crop, batch_ids=None, **kwargs):  # pragma: no cover
+    return gen_cluster_script(crop, 'slurm', batch_ids=batch_ids, **kwargs)
+
+
+gen_slurm_script.__doc__ = \
+    _specific_scheduler_gen_cluster_script_doc.format(scheduler='slurm')
+
+Crop.gen_sge_script = gen_sge_script
+Crop.gen_pbs_script = gen_pbs_script
+Crop.gen_slurm_script = gen_slurm_script
+
+_specific_scheduler_grow_cluster_doc = \
+    """Automagically submit {scheduler} jobs to grow all missing results.
+
+    Parameters
+    ----------
+    crop : Crop
+        The crop to grow.
+    batch_ids : int or tuple[int]
+        Which batch numbers to grow, defaults to all missing batches.
+    kwargs
+        See `grow_cluster` for all other parameters (except for
+        `scheduler` which is not needed here).
+    """
+
+
+def grow_sge(crop, batch_ids=None, **kwargs):  # pragma: no cover
+    grow_cluster(crop, 'sge', batch_ids=batch_ids, **kwargs)
+
```

Comment:
I thought about doing this, but didn't for 2 reaons:

1) I think giving `scheduler` to `partial` as a keyword cause an error if `batch_ids` is given as a positional arg?

2) The docstring will be wrong, as it will include the `scheduler` arg

Reply:
1. I don't think so (and have just tried a toy example), think its just like calling ``fn(*arg, **kwargs, scheduler='sge')``
2. The docs function signature should show the arg has already been set, I think better to have one extra arg like this in the docstring than having to look up most args in a different function.

```python
In [1]: import functools

In [2]: def foo(a, *, b):
    ...:     """Foo docs.
    ...:     """
    ...:     return a + b
    ...:

In [3]: foo2 = functools.partial(foo, b=2)

In [4]: functools.update_wrapper(foo2, foo)
Out[4]: functools.partial(<function foo at 0x7fda74eab950>, b=2)

In [5]: foo2(1)
Out[5]: 3

In [6]: foo2?
Signature:       foo2(a, *, b)
Call signature:  foo2(*args, **kwargs)
Type:            partial
String form:     functools.partial(<function foo at 0x7fda74eab950>, b=2)
File:            ~/<ipython-input-30-8b71f7e4af39>
Docstring:
Foo docs.

Class docstring:
partial(func, *args, **keywords) - new function with partial application
of the given arguments and keywords.
```

Reply:
I was convinced - I tried
```python
grow_slurm = functools.partial(grow_cluster, scheduler='slurm')
functools.update_wrapper(grow_slurm, grow_cluster)
Crop.grow_slurm = grow_slurm
```
but when I tried to call it with 
```python
c = xyz.Crop(name=f'mb_32', parent_dir='/ddn/data/fscs35/xyzpy_m2sbnb2')
c.grow_slurm(batch_ids=(2,), hours=24, gigabytes=4, temp_gigabytes=0) 
```
I get
```python
TypeError: grow_cluster() missing 1 required positional argument: 'crop'
```
It seems that assigning a `partial` object as an attribute to a class doesn't make it get treated like an instance method in the way that it does with a function.

Reply:
mm this is unknown territory for me. Maybe it needs to be a separate ``functools.partialmethod`` wrap? I've always wondered why this was needed. E.g.
```python
Crop.grow_slurm = functools.partialmethod(Crop.grow_cluster, scheduler='slurm')
```

Reply:
Cool, I've not seen that before, but from reading the docs it sounds like it should work. I'll try tomorrow.

Reply:
It seems to on a simple ex:
```python
import functools

class Foo:
     def __init__(self, x):
         self.x = x

def add(f, a, b):
    return f.x + a + b


Foo.add = add
Foo.add3 = functools.partialmethod(Foo.add, b=3)

f = Foo(1)
f.add(2, 3) == f.add3(2) == 6
# True
```
thanks, sorry this has ended up being more complex than I thought!

Reply:
After some digging, I've found out that the reason it doesn't work is because, unlike `function`, `partial` doesn't adhere to the 'descriptor protocol' which allows a `function` attribute of instances to turn into an `instancemethod` object when its name is resolved. The `partialmethod` object does implement the protocol, which means that `partialmethod` objects get turned into `partial` objects when the name gets resolved.

Unfortunately, `partialmethod` objects _must_ be used as methods: they can't be directly called which means we'd have to make a `partial` version for the module-level functions and `partialmethod` versions for the `Crop` methods. Also, `update_wrapper` seems to do nothing on `partialmethod` objects - at least, I've tried all variations I can think of and it doesn't seem to work.

A much less complicated thing would be to do this:
```python
def grow_slurm(*args, **kwargs):
    return grow_cluster(*args, **{**{'scheduler':'slurm'}, **kwargs})
functools.update_wrapper(grow_slurm, grow_cluster)
Crop.grow_slurm = grow_slurm
```
which works fine for the module-level version and the `Crop` method, including the updated wrapper, but the docs won't give any hint that one argument has been fixed.

What do you think?

Reply:
The plot thickens! I do think it should be pretty clear from the context that scheduler has been set... but I also think maybe the answer is just to remove the module level functions, and only have methods / partialmethods.